rfd/0073-kube-secret-storage.md

```diff
@@ -0,0 +1,501 @@
+
+## Details
+
+### Limitations
+
+High availability can only be achieved using Statefulsets and not deployments. This means that if the desired number of replicas is bigger than one, i.e `replicas>1`, the [Helm chart](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-kube-agent) has to switch to Statefulset objects instead of Deployments. This change is required because some Kubernetes invariant must be kept between restarts in order to correctly map each agent pod to its identity stored in the secret. The invariant used is the Statefulset pod names `{{ .Release.Name }}-{0...replicas}}`.
+
+Given this, it is required to expose the `$TELEPORT_REPLICA_NAME` environment variable to each pod in order to the backend storage be able to write the identity separately. The values for `$TELEPORT_REPLICA_NAME` are dependent on the object type:
+
+- Deployment: `TELEPORT_REPLICA_NAME` is a constant string `{{ .Release.Name }}`.
+- Statefulset: `TELEPORT_REPLICA_NAME` is a dynamic value provided by Kubernetes `fieldPath: metadata.name`.
+
+If for some reason the operator increase the number of replicas after deploying the agent with a Deployment object, all agents will use the exact same identity when accessing the cluster.
+
+### Secret creation and lifecycle
+
+In order to use the secret storage, the secret can be created by [Helm chart](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-kube-agent) at install time. This is a requirement for the cleanup procedures once the user run `helm uninstall {release}`.
+
+In order to support the secret creation via Helm, the Helm chart has to be changed to reuse the Secret data if it exists in Kubernetes and, if does not exist, create an empty secret. This can be achieved with the Helm's native function `lookup`.
+
+```helm
+{{- $secret := lookup "v1" "Secret" .Release.Namespace "{{ .Release.Name }}-identity" -}}
+{{- if $secret -}}
+{{/*
+   Reusing secret data
+*/}}
+{{- else -}}
+{{/*
+    create secret with empty data
+*/}}
+{{- end -}}
+
+```
+
+#### Secret content
+
+Once Kube Agent updates the secret, it will have the following structure:
+
+```yaml
+apiVersion: v1
+data: |
+    {{$TELEPORT_REPLICA_NAME}}:
+        {{.Values.kubeClusterName}}:
+            {
+            "kind": "identity",
+            "version": "v2",
+            "metadata": {
+                "name": "current"
+            },
+            "spec": {
+                "key": "{key_content}",
+                "ssh_cert": "{ssh_cert_content}",
+                "tls_cert": "{tls_cert_content}",
+                "tls_ca_certs": ["tls_ca_certs"],
+                "ssh_ca_certs": ["ssh_ca_certs"]
+            }
+            }
+// if running in high availabilty mode
+    {{$TELEPORT_REPLICA_NAME[i]}}:
```

Comment:
Do I understand correctly that this is including multiple identities in the same Secret? Why not save each identity into its own Secret? This would be better from RBAC perspective too.

Reply:
I separated them for readability